Discover GitHub tokens stored by the GitHub CLI

neersighted opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Introduce the ability to discover tokens stored by the GitHub CLI. The changes to the `GitHub::API` are relatively straightforward; the current implementation will only discover the `gh` command from the `gh` formula.

neersighted commented on Library/Homebrew/utils/github/api.rb:
```diff
@@ -65,16 +65,23 @@ class AuthenticationFailedError < Error
+          <<~EOS
+            HOMEBREW_GITHUB_API_TOKEN may be invalid or expired; check:
+              #{Formatter.url("https://github.com/settings/tokens")}
             #{GitHub.pat_blurb}
```
I moved the `pat_blurb` here as I don't think it really makes sense with `keychain_username_password` -- instructions on creating a PAT and assiging it to `HOMEBREW_GITHUB_API_TOKEN` make more sense in the context of that token.

If a user has an invalid set of creds in the keychain, they will first be prompted to erase them. Then, if they still need credentials (e.g. for a private repo), they'll get clear instructions on how to create the token. This way the keychain functionality remains largely in the background.

MikeMcQuaid replied:
Will this not now be output even if `Homebrew::EnvConfig.github_api_token` is unset? If so, that's undesirable.

neersighted replied:
The prior behavior was to complain of invalid credentials in the keychain and suggest creating a PAT and storing it in `HOMEBREW_GITHUB_API_TOKEN`; the fact that the keychain was referred to at all is unexpected for most users I think. Is this not a bit more consistent?

MikeMcQuaid replied:
I think an extra case checking if `if Homebrew::EnvConfig.github_api_token.present?` and saying `HOMEBREW_GITHUB_API_TOKEN is unset` or similar would improve this, thanks!

carlocab commented on Library/Homebrew/utils/github/api.rb:
```diff
@@ -129,29 +152,31 @@ def self.keychain_username_password
 
     def self.credentials
-      @credentials ||= Homebrew::EnvConfig.github_api_token || keychain_username_password
+      @credentials ||= Homebrew::EnvConfig.github_api_token || github_cli_token || keychain_username_password
```
Might be safer to put `github_cli_token` last here, to avoid changing behaviour for users who expect `brew` to check the Keychain.

MikeMcQuaid replied:
Agreed 👍🏻

neersighted replied:
Hmm, I think I would prefer to invert this behavior as a follow-up I had in mind was to make use of **any** `credential-helper` the user configured by using `git credential fill`. I'm happy to swap the order back here, but in the longer run I think it makes more sense to prefer the GitHub CLI over Git credential helpers for a myriad of reasons (including that the GitHub CLI **may be** the credential helper and it avoids triggering GUI prompts/is a 'better' source).

MikeMcQuaid replied:
People have historically been fairly twitchy (understandably) about us changing the order here so I think it should wait for a major/minor release so we can put it in the release notes.

```suggestion
      # This is not odeprecated but use that magic string so we put
      # github_cli_token before keychain_username_password on next major/minor release
      @credentials ||= Homebrew::EnvConfig.github_api_token || keychain_username_password || github_cli_token
```

should do it.

carlocab replied:
Yup, happy to switch up the order at some point, but it should wait till we can announce it in release notes.

carlocab reviewed (approved):
Looks reasonable to me!

carlocab commented:
Test failure fixed by #15486.

MikeMcQuaid commented on Library/Homebrew/utils/github/api.rb:
```diff
@@ -112,15 +119,31 @@ def initialize(github_message, errors)
+    sig { returns(T.nilable(String)) }
+    def self.github_cli_token
+      env = { "PATH" => PATH.new(Formula["gh"].opt_bin, ENV.fetch("PATH")) }
+      out, _, result = system_command "gh",
```
```suggestion
      gh_output, _, result = system_command "gh",
```

MikeMcQuaid commented on Library/Homebrew/utils/github/api.rb:
```diff
@@ -112,15 +119,31 @@ def initialize(github_message, errors)
-      end
-      github_username = github_credentials[/username=(.+)/, 1]
-      github_password = github_credentials[/password=(.+)/, 1]
+      out, _, result = system_command "git",
```
```suggestion
      git_output, _, result = system_command "git",
```

MikeMcQuaid reviewed (commented):
Wow, really great work so far, lovely to receive a PR like this out of the blue! A few nits but should be good to merge soon. Thanks!

neersighted commented:
I've pushed up a new set of commits addressing review comments. I think this should be acceptable to everyone for now, and there's a bit more in the way of improvements I want to make in this area, but they're tangentially related/more controversial, so let's save those for a second PR.

carlocab commented on Library/Homebrew/utils/github/api.rb:
```diff
@@ -62,19 +62,30 @@ def pretty_ratelimit_reset(reset)
-              #{Formatter.url("https://github.com/settings/tokens")}
+            Your GitHub CLI login session may be invalid.
+            Refresh it with:
+              gh auth login -h github.com
```
```suggestion
              gh auth login --hostname github.com
```

carlocab reviewed (approved):
Thanks!

carlocab commented on Library/Homebrew/utils/github/api.rb:
```diff
@@ -62,19 +62,30 @@ def pretty_ratelimit_reset(reset)
+          EOS
+        else
+          <<~EOS
+            HOMEBREW_GITHUB_API_TOKEN is unset.
```
Yes, I think so. Maybe we just need to throw `MissingAuthenticationError` instead if we fall through to here?

neersighted replied:
The case where this could be hit is if you try to take an action that requires authentication against a publicly visible resource. I'm not sure how often Homebrew might attempt that, but it is worth nothing that `test-bot` was attempting that before https://github.com/Homebrew/brew/pull/15486.

In any case, https://github.com/Homebrew/brew/pull/15485#discussion_r1206425367 was the driver behind adding this; I'm happy to walk it back if consensus changes.

neersighted replied:
Combing over the API docs a bit, it looks like there are only a couple scenarios we should run into:
* [401 when the credentials are invalid](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#failed-login-limit)
* [403 when no User-Agent is provided](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#user-agent-required)
* [403 when not enough scope is available](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#authentication)
* [403 when we go over the ratelimit](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit)

As the only scenarios outside rate-limiting where we can hit 401/403 happen with credentials, I'll go ahead and re-work this code a bit.

neersighted replied:
In the latest iteration I've just let this case fall through as it doesn't seem like it should happen in the 'real world'; I've also left the `pat_blurb` out for now as it doesn't seem to provide any further useful context when the token is invalid.

If the user clears their token and then hits an auth error again, they will get the PAT blurb which is now more actionable/meaningful.

MikeMcQuaid commented on Library/Homebrew/utils/github/api.rb:
```diff
@@ -62,30 +62,36 @@ def pretty_ratelimit_reset(reset)
     class MissingAuthenticationError < Error
       def initialize
-        message = +"No GitHub credentials found in macOS Keychain or environment.\n"
+        message = +"No GitHub credentials found via macOS keychain, GitHub CLI, or the environment.\n"
```
```suggestion
        message = +"No GitHub credentials found in macOS Keychain, GitHub CLI or the environment.\n"
```

MikeMcQuaid reviewed (commented):
Looking good!

MikeMcQuaid commented:
Thanks so much for your first contribution to Homebrew (hopefully of many)! Without people like you submitting PRs we couldn't run this project. You rock, @neersighted!
